xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/test/java/org/xwiki/rest/internal/ModelFactoryTest.java

```diff
@@ -0,0 +1,142 @@
+    {
+        when(xcontextProvider.get()).thenReturn(xcontext);
+        
+        BaseObject xwikiObject = mock(BaseObject.class);
+        BaseClass xwikiClass  = mock(BaseClass.class);
+       
+        when(xwikiObject.getPropertyNames()).thenReturn(new String[] {});
+        when(xwikiObject.getXClass(xcontext)).thenReturn(xwikiClass);
+        when(xwikiObject.getClassName()).thenReturn("Some.XClass");
+        when(xwikiObject.getNumber()).thenReturn(0);
+        
+        StringClass stringField = new StringClass();
+        stringField.setName("textValue");
+        StringProperty textElement = new StringProperty();
+        textElement.setName("textValue");
+        textElement.setClassType("String");
+        textElement.setValue("abcd");
+        when(xwikiObject.get("textValue")).thenReturn(textElement);
+
+        PasswordClass pwField = new PasswordClass();
+        pwField.setName("passwordValue");
+        StringProperty pwElement = new StringProperty();
+        pwElement.setName("passwordValue");
+        pwElement.setClassType("Password");
+        pwElement.setValue("abcd");
+        when(xwikiObject.get("passwordValue")).thenReturn(pwElement);
+        
+        when(xwikiClass.getProperties()).thenReturn(new java.lang.Object[] {stringField, pwField});
+        
+        // some more mockery needed to pass the "toRestObject" call
+        // although the results are not tested (yet)
+
+        URI dummy = new URI("https://localhost/");
+        Document doc = mock(Document.class);
+        SpaceReference spaceRef = new SpaceReference("Space", new WikiReference("wiki"));
+        when(doc.getPrefixedFullName()).thenReturn("wiki:Space.Page");
+        when(doc.getWiki()).thenReturn("wiki");
+        when(doc.getSpace()).thenReturn("Space");
+        when(doc.getName()).thenReturn("Page");
+        when(doc.getDocumentReference()).thenReturn(new DocumentReference("Page", spaceRef));
+        when(resolver.resolve(eq("Space"), any(String.class))).thenReturn(spaceRef);
+        
+        com.xpn.xwiki.web.Utils.setComponentManager(componentManager);
+        
+        //
+        // now we get around to call the method under test
+        //
+        
+        Object result = modelFactory.toRestObject(dummy, doc, xwikiObject, false, false);
+
+        List<Property> properties = result.getProperties();
+        // note: we do not mind if there is no "Password" property at all in the list
+        // that is as good as not showing the value
+        boolean stringFieldFound = false;
+        for (Property prop : properties) {
+            if ("String".equals(prop.getType())) {
+                assertEquals("abcd", prop.getValue(), "should see proper value for StringField");
+                stringFieldFound = true;
+            }
+            if ("Password".equals(prop.getType())) {
+                assertNull(prop.getValue(), "should see not value for PasswordField");
+            }
```

Comment:
So we want allow Password types but they must have null values? Or should there be no Password property? I'm asking since it's not clear from the test what is the contract we want.

Reply:
From What I see at PageResourceTest, ideally we should add a new test method there like `testPUTGETWithPasswordXProperty` or modify the existing `testPUTGETWithObject` but instead of using the Tag XObject, use a XWikiUsers xobjects for example (which contains a password field).

You test verifies that we don't show the password which is cool but it doesn't test that we can set it. Doing it the way it's done in PageResourceTest would validate both aspects (PUT and GET).

Note to self: we need to convert PageResourceTest to a JUnit5 test.

@ClemensRobbenhaar Don't mistake me. I'm very happy that you wrote the fix + the test. For me that's good enough to close the issue.

Reply:
It checks that if we have a Password at all, then its value will not be exposed.
The current implementation just skips the ``<value>`` element from the output; i.e. REST Clients still can see that a user has a password, but nit which value it has.
This is not an intentional decision on my side, it was just the quickest way to fix things over the weekend.